lib/FIR/statements.h

```diff
@@ -265,12 +285,15 @@ class ActionStmt_impl : virtual public Stmt_impl {
 class ApplyExprStmt : public ActionStmt_impl {
 public:
   static ApplyExprStmt Create(const Expression *e) { return ApplyExprStmt{*e}; }
-  static ApplyExprStmt Create(Expression &&e) { return ApplyExprStmt{e}; }
+  static ApplyExprStmt Create(Expression &&e) {
+    return ApplyExprStmt{std::move(e)};
+  }
 
   Expression expression() const { return expression_; }
 
 private:
   explicit ApplyExprStmt(const Expression &e) : expression_{e} {}
+  explicit ApplyExprStmt(Expression &&e) : expression_{e} {}
```

Comment:
done

Reply:
github won't let me reply to the general comment, so I'll put my reply here.

> I think it may be worth improving the fact that a `QualifiedSmt` is actually containing a pointer to a Statement of the right variant,

I am not sure I'm following.  A `QualifiedStmt` is ultimately (after type erasure) nothing but a pointer.  Or if you rather, it is a pointer with some suitable type wrappings for the C++ type system that designate precisely: this wrapped `Statement` does/must/will always be a member of the given set of subtypes.
 
> and I do not know if you can completely get rid of dynamic checks if Statements are mutable. 

They are not.  They are instantiated once.  The intuition is that for a given operation, there can be any number of other operations that use the result.  Morphing that operation into something else arbitrarily will almost surely violate the overarching semantics of the program being represented, since that would imply changing the operation universally in all contexts.

> But I like the idea of having the static representation of what should be inside a Statement.

It's worth pointing out that we've always had the ability to recover what the exact statement implementation type of any generic `Statement` is; i.e., `Statements` have always used parametric polymorphism.  We've also had sets of subtypes for subtype polymorphism, and these sets are built into the implementation type hierarchy.  This layer gives control over subtype constraints.

By analogy, a basic block is a container ship.  Everything shipped to Hamburg goes in rectangular containers (statements) on the ship.  We've know how to open any container and see exactly what's inside it (the exact statement variant).  This gives us the property, whatever is inside a particular container is known to be blue, whether it is a car, a keyboard, or a taco.